plugins/teststeps/teststeps.go

```diff
@@ -27,53 +29,80 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
-	for {
-		select {
-		case target := <-ch.In:
-			if target == nil {
-				// no more targets incoming
-				return nil
+	type tgtErr struct {
+		target *target.Target
+		err    error
+	}
+	tgtErrCh := make(chan tgtErr)
+	tgtInFlight := int32(0)
+	noMoreTargetsCh := make(chan struct{})
+
+	go func(tgtInFlight *int32) {
+		defer func() {
+			log.Debugf("%s: ForEachTarget: exiting incoming loop, targets inflight %d", pluginName, tgtInFlight)
+		}()
+		for {
+			tgt := <-ch.In
+			log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
+
+			if tgt == nil {
+				log.Debugf("%s: ForEachTarget: all targets have been received", pluginName)
+				noMoreTargetsCh <- struct{}{}
+				return
 			}
-			log.Debugf("%s: ForEachTarget: received target %s", pluginName, target)
-			errCh := make(chan error)
+
 			go func() {
-				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, target)
-				errCh <- f(cancel, pause, target)
+				tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
 			}()
+			atomic.AddInt32(tgtInFlight, 1)
+		}
+	}(&tgtInFlight)
 
-			select {
-			case err := <-errCh:
-				if err != nil {
-					select {
-					case ch.Err <- cerrors.TargetError{Target: target, Err: err}:
-						log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, target, err)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
-				} else {
-					select {
-					case ch.Out <- target:
-						log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, target)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
+	noMoreTargets := false
+	defer func() {
+		log.Debugf("%s: ForEachTarget: exiting outgoing loop, targets inflight %d, the last target received %v", pluginName, tgtInFlight, noMoreTargets)
+	}()
+	for {
+		select {
+		case te := <-tgtErrCh:
+			atomic.AddInt32(&tgtInFlight, -1)
+			if te.err != nil {
+				select {
+				case ch.Err <- cerrors.TargetError{Target: te.target, Err: te.err}:
```

Comment:
I am hoping to simplify this interface with https://github.com/facebookincubator/contest/pull/121, so that there will be only a unified channel, where an error might be nil. Having two channels requires always having an else branch, which makes things significantly more difficult to read and to maintain.

Reply:
It would be nice to use simplified channels here, but does it mean PR should wait for #121 to be accepted?

Reply:
Hehe, much worse #121 to be accepted, which depends on #119, which depends #118, which depends on having a tool to do schema migrations. Lots of work to unblock #121 still to do, and I don't foresee to be able to get it done in the short term. This should definitely go out before #121, it was mostly an FYI.